src/lua/init.lua

```diff
@@ -203,7 +203,31 @@ rawset(package, "search", search)
 rawset(package, "searchroot", searchroot)
 rawset(package, "setsearchroot", setsearchroot)
 
+-- Execute scripts or load modules pointed by TT_PRELOAD
+-- environment variable.
+local function run_preload()
+    local tt_preload = os.getenv('TT_PRELOAD') or ''
+    assert(type(tt_preload) == 'string')
+    if #tt_preload == 0 then
+        return
+    end
+    for _, script in ipairs(tt_preload:split(';')) do
```

Comment:
Side note: Can (and should) we support the format with `;` at the end:

```sh
$ TT_PRELOAD="src/test.lua;src.test2;" src/tarantool -e 'print(1)'
0       src/tarantool
-1      /home/burii/reviews/tarantool/Totktonada/src/tarantool
src/test.lua
0       src/tarantool
-1      /home/burii/reviews/tarantool/Totktonada/src/tarantool
src.test2
LuajitError: builtin/tarantool.lua:222: module '' not found:
        no field package.preload['']
        no file '/home/burii/reviews/tarantool/Totktonada/.lua'
        no file '/home/burii/reviews/tarantool/Totktonada//init.lua'
        no file '/home/burii/reviews/tarantool/Totktonada/.so'
        no file '/home/burii/reviews/tarantool/Totktonada/.rocks/share/tarantool/.lua'
        no file '/home/burii/reviews/tarantool/Totktonada/.rocks/share/tarantool//init.lua'
        no file '/home/burii/reviews/tarantool/.rocks/share/tarantool/.lua'
        no file '/home/burii/re
fatal error, exiting the event loop
```

Maybe we can just skip empty `''` entries?

Reply:
We definitely can. Should we? Don't know.

This idea comes to me as well. I didn't find why one may request it.

Now I think a bit more and find that it may ease construction of this variable using concatenation. So maybe it worth it.

I'll raise the same discussion in the internal RFC document and will come here with updates.

Reply:
Quote from the RFC:

> ### Whether to allow empty entries?
>
> From @Buristan and @Totktonada.
>
> If we’ll go with the direct solution (`:split(';')` without filtering empty entries), the behavior will be the following:
>
> ```bash
> TT_PRELOAD='foo;bar' tarantool  # correct: load 'foo' and 'bar'
> TT_PRELOAD='foo;bar;' tarantool # error
> TT_PRELOAD='foo;;bar' tarantool # error
> ```
>
> We can allow empty entries to ease construction of the environment valuable value using concatenation.
>
> This way it'll work in the following way:
>
> ```bash
> TT_PRELOAD='foo;bar' tarantool  # correct: load 'foo' and 'bar'
> TT_PRELOAD='foo;bar;' tarantool # correct: load 'foo' and 'bar'
> TT_PRELOAD='foo;;bar' tarantool # correct: load 'foo' and 'bar'
> ```
>
> Decision: allow.

@igormunkin's comment:

> Just consider empty string/path/module name as a noop. Hence there will be no issues with multiple/trailing separators. The only weak point of such approach is validation, but I doubt we have to validate empty argument in any other way than ignoring it.

@unera agreed too.

I'll update the patch.

Reply:
<details>
<summary>The fixup</summary>

```diff
diff --git a/src/lua/init.lua b/src/lua/init.lua
index 723d16f44..77072e7c2 100644
--- a/src/lua/init.lua
+++ b/src/lua/init.lua
@@ -212,7 +212,13 @@ local function run_preload()
         return
     end
     for _, script in ipairs(tt_preload:split(';')) do
-        if script:endswith('.lua') then
+        if #script == 0 then
+            -- Ignore empty entries to allow duplicated semicolons
+            -- and leading/trailing semicolons.
+            --
+            -- It simplifies construction of the environment
+            -- variable value using concatenation.
+        elseif script:endswith('.lua') then
             local fn, err = loadfile(script)
             if fn == nil then
                 error(err)
diff --git a/test/app-luatest/preload_test.lua b/test/app-luatest/preload_test.lua
index c924f5866..cd8ba3d07 100644
--- a/test/app-luatest/preload_test.lua
+++ b/test/app-luatest/preload_test.lua
@@ -184,7 +184,10 @@ local function expected_output(scripts, env, args)
     -- TT_PRELOAD entries.
     for _, entry in ipairs(env['TT_PRELOAD']:split(';')) do
         local script
-        if entry:endswith('.lua') then
+        if #entry == 0 then
+            -- Ignore empty entries.
+            goto continue
+        elseif entry:endswith('.lua') then
             -- The entry is a script file.
             script = entry
         else
@@ -219,6 +222,8 @@ local function expected_output(scripts, env, args)
             ['arg[0]'] = args[1],
             ['arg[]'] = script_args,
         })
+
+        ::continue::
     end
 
     -- The main script.
@@ -296,6 +301,30 @@ for _, case in ipairs({
         env = {TT_PRELOAD = 'foo.lua;bar'},
         args = {'main.lua'},
     },
+    {
+        'test_leading_semicolon',
+        scripts = {'foo.lua', 'bar.lua', 'main.lua'},
+        env = {TT_PRELOAD = ';foo.lua;bar'},
+        args = {'main.lua'},
+    },
+    {
+        'test_trailing_semicolon',
+        scripts = {'foo.lua', 'bar.lua', 'main.lua'},
+        env = {TT_PRELOAD = 'foo.lua;bar;'},
+        args = {'main.lua'},
+    },
+    {
+        'test_duplicated_semicolon',
+        scripts = {'foo.lua', 'bar.lua', 'main.lua'},
+        env = {TT_PRELOAD = 'foo.lua;;bar'},
+        args = {'main.lua'},
+    },
+    {
+        'test_only_semicolons',
+        scripts = {'main.lua'},
+        env = {TT_PRELOAD = ';;;'},
+        args = {'main.lua'},
+    },
     -- Negative cases.
     {
         'test_nonexisting_script',
```
</details>

Updated the commit/PR description this way:

> Redundant (duplicated, leading, trailing) semicolons are ignored.

Reply:
Ouch. Fixup of the fixup:

```diff
diff --git a/src/lua/init.lua b/src/lua/init.lua
index 77072e7c2..7f8b7c143 100644
--- a/src/lua/init.lua
+++ b/src/lua/init.lua
@@ -212,6 +212,7 @@ local function run_preload()
         return
     end
     for _, script in ipairs(tt_preload:split(';')) do
+        -- luacheck: ignore 542 empty if branch
         if #script == 0 then
             -- Ignore empty entries to allow duplicated semicolons
             -- and leading/trailing semicolons.
```